Bootintobios

DaanHoogland opened the pull request:
UI implementation for https://github.com/apache/cloudstack/pull/4021

DaanHoogland commented:
@rhtyd @shwstppr can you review please?

rhtyd commented on src/config/section/compute.js:
```diff
@@ -86,7 +86,7 @@ export default {
           groupAction: true,
           show: (record) => { return ['Stopped'].includes(record.state) },
-          args: ['podid', 'clusterid', 'hostid'],
+          args: ['podid', 'clusterid', 'hostid', 'bootintobios'],
```
@DaanHoogland is this feature allowed for users as well?

DaanHoogland replied:
yes as i understand it it is

rhtyd commented on src/views/compute/DeployVM.vue:
```diff
@@ -275,6 +275,16 @@
                 </template>
               </a-step>
+              <a-step
+                :title="this.$t('enter BIOS setup')"
```
`this.$t..` can be written as `$t(...). Instead of a string, can you use a key, we usually encourage the same key as the API arg/parameter if applicable. You can then define the key and the language translation in src/locales/en.json

rhtyd commented on src/views/compute/DeployVM.vue:
```diff
@@ -275,6 +275,16 @@
+                <template slot="description">
+                  <div v-if="zoneSelected">
+                    <a-checkbox
+                      @change="enterBiosChoiceChanged">enter BIOS setup</a-checkbox>
```
if it's just a setter, we can also write shorthand like this:
```
@change="val => bootIntoBios = val"
```

DaanHoogland replied:
i'll test this, i didn't get any other way to work, yet.

rhtyd commented on src/views/compute/DeployVM.vue:
```diff
@@ -918,6 +933,8 @@ export default {
         deployVmData.name = values.name
         deployVmData.displayname = values.name
+        // step 8: enter setup
+        deployVmData.bootintobios = this.bootintobios
```
I think this setting can be at the top step itself where we're getting name, keyboard, etc.

DaanHoogland replied:
ok, kinda makes sense

shwstppr commented on src/views/compute/DeployVM.vue:
```diff
@@ -275,6 +275,16 @@
+                <template slot="description">
+                  <div v-if="zoneSelected">
+                    <a-checkbox
+                      @change="enterBiosChoiceChanged">enter BIOS setup</a-checkbox>
```
```suggestion
                      @change="event => { this.bootintobios = event.target.checked }">enter BIOS setup</a-checkbox>
```
@DaanHoogland

DaanHoogland commented:
thanks guys, I'll address your feedback, soon.

DaanHoogland commented on src/config/section/compute.js:
```diff
@@ -86,7 +86,7 @@ export default {
           groupAction: true,
           show: (record) => { return ['Stopped'].includes(record.state) },
-          args: ['podid', 'clusterid', 'hostid'],
+          args: ['podid', 'clusterid', 'hostid', 'bootintobios'],
```
last args should be conditional on `hypervisor === vmware`. will be done after #231 has been merged with its [AutoGenView.vue changes](https://github.com/apache/cloudstack-primate/pull/231/files#diff-83f2f8976f46b2444d31fcd1895d8165R548)

DaanHoogland commented on src/config/section/compute.js:
```diff
@@ -104,7 +104,8 @@ export default {
           dataView: true,
-          show: (record) => { return ['Running'].includes(record.state) }
+          show: (record) => { return ['Running'].includes(record.state) },
+          args: ['bootintobios']
```
this args field should be conditional on `hypervisor === vmware`. will be done after #231 has been merged with its [AutoGenView.vue changes](https://github.com/apache/cloudstack-primate/pull/231/files#diff-83f2f8976f46b2444d31fcd1895d8165R548)

DaanHoogland commented:
thanks @rhtyd and @shwstppr . I addressed your comments, please have a look. And please note that this code will at best do nothing until [cloudstack 4021](https://github.com/apache/cloudstack/pull/4021) is merged.

DaanHoogland commented on src/config/section/compute.js:
```diff
@@ -104,7 +104,8 @@ export default {
           dataView: true,
-          show: (record) => { return ['Running'].includes(record.state) }
+          show: (record) => { return ['Running'].includes(record.state) },
+          args: ['bootintosetup']
```
@shwstppr , I should be able to replace this with
```
args: (record) -> { return record.hypervisorType === 'vmware' ? ['bootintosetup'] : [] }
```
am I? It seems not to work. How can I find what to use as parameter of the anonymous function and what attributes it has?

tnx

shwstppr commented on src/config/section/compute.js:
```diff
@@ -104,7 +104,8 @@ export default {
           dataView: true,
-          show: (record) => { return ['Running'].includes(record.state) }
+          show: (record) => { return ['Running'].includes(record.state) },
+          args: ['bootintosetup']
```
```suggestion
          args: (record) => { return record.hypervisor === 'VMware' ? ['bootintosetup'] : [] }
```
@DaanHoogland 
This should work but currently it seems record is coming undefined from AutogenView. I've created PR #327 to fix that.
`record` here will be the response for `listvirtualmachinesmetrics` response item

DaanHoogland replied:
ah to bad, I want to know the hypervisor type for the machine. @shwstppr any tips?

shwstppr replied:
@DaanHoogland `record.hypervisor` with give you that.

VMs in primate are listed with listvirtualmachinesmetrics API and it only has hypervisor param. No parameter named hypervisortype in API response.

DaanHoogland replied:
hypervisor contains the type, tnx. There is something confusing in the API doc, it only lists it as a request parameter. I think I should be fine.

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build primate packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ?centos ?debian ?archive. JID-1863

blueorangutan commented:
Packaging result: ✔centos ✔debian ✔archive. JID-

blueorangutan commented:
Packaging result: ✔centos ✔debian ✔archive. JID-

blueorangutan commented:
Packaging result: ?centos ?debian ?archive. JID-1864

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build primate packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark:centos :heavy_check_mark:debian :heavy_check_mark:archive. JID-1869

rhtyd commented:
@DaanHoogland please ping me when this is ready for review

DaanHoogland commented:
@rhtyd it is ready for review, but the cloudstack code is not merged yet.

rhtyd commented:
@DaanHoogland can you fix the merge conflicts? I think we merge this in Primate as soon as we've any test confirmation. Thanks.

DaanHoogland commented:
@shwstppr @davidjumani @rhtyd I rebased but my en.json file does not seem ok. to much diff. Can you check if I did the reverse of what was intended?

rhtyd commented:
@DaanHoogland not urgent - the next time you get chance to work on this - can you revert your changes to en.json; keep the version from upstream and then sort/diff what you need to specific keys for your feature.

DaanHoogland commented:
@rhtyd fixed my booboo (hope the late mister Hawkins approves)

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build primate packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark:centos :heavy_check_mark:debian :heavy_check_mark:archive.
QA: http://primate-qa.cloudstack.cloud:8080/client/pr/313 (JID-1978)

rhtyd commented:
@borisstoyanov @vladimirpetrov have you also tested this Primate PR against the upstream PR (merged upstream now) https://github.com/apache/cloudstack/pull/4021?

borisstoyanov commented:
@rhtyd yes, primate changes has been covered also

rhtyd commented on src/config/section/compute.js:
```diff
@@ -89,7 +89,18 @@ export default {
+              fields = ['podid', 'clusterid', 'hostid']
+            }
+            if (record.hypervisor === 'VMware') {
+              if (store.apis.startVirtualMachine.params.filter(x => x.name === 'bootintosetup').length > 0) {
```
@DaanHoogland I've tested and fixed a case here - we want to make Primate work with all recent/pass/future version of CloudStack; so if there's a feature adding new params for autogenerated forms, we can check if the param is available by checking the API discovery /cc @shwstppr @davidjumani
